[WIP] Gradient Boosting binding (#3735)
mlpack/mlpack: doc/user/methods/gradient_boosting.md

## `GradBoosting`

The `GradBoosting` class implements a Gradient Boosting model that supports
numerical and categorical features. The class offers several template 
parameters and several runtime options that can be used to control the 
model.

Gradient Boosting is a very powerful ensemble algorithm used for both 
classification and regression tasks. It utilizes a series of weak learners 

> (eg. [Decision Tree Regressor](decision_tree_regressor.md)) to arrive closer and closer to a targeted label. Each 

Reviewer: You used `e.g.` here which means "for example" or "for instance", but in the implementation is hard-coded to use `DecisionTreeRegressor`.  So I would instead suggest:

```suggestion
(specifically [decision tree regressors](decision_tree_regressor.md)) to arrive closer and closer to a targeted label. Each 
```

arma::mat dataset(10, 1000, arma::fill::randu); // 1000 points.
arma::Row<size_t> labels =
    arma::randi<arma::Row<size_t>>(1000, arma::distr_param(0, 4));
arma::mat testDataset(10, 500, arma::fill::randu); // 500 test points.
size_t numWeakLearners = 10;
size_t numClasses = 5;

mlpack::GradBoosting gb;               // Step 1: create model.
gb.Train(dataset, labels, numClasses, numWeakLearners);          // Step 2: train model.
arma::Row<size_t> predictions;
gb.Classify(testDataset, predictions); // Step 3: classify points.

// Print some information about the test predictions.
std::cout << arma::accu(predictions == 2) << " test points classified as class "
    << "2." << std::endl;
```
<p style="text-align: center; font-size: 85%"><a href="#simple-examples">More examples...</a></p>

#### Quick links:

 * [Constructors](#constructors): create `GradientBoosting` objects.
 * [`Train()`](#training): train model.
 * [`Classify()`](#classification): classify with a trained model.
 * [Other functionality](#other-functionality) for loading, saving, and
   inspecting.
 * [Examples](#simple-examples) of simple usage and links to detailed example
   projects.

#### See also:

>  * [`ADABoost`](adaboost.md)

Reviewer: ```suggestion
 * [`AdaBoost`](adaboost.md)
```
(just fixing the name of the C++ class)

| _multi-point_ | `predictedLabels` | [`arma::Row<size_t>&`](../matrices.md) | Vector of `size_t`s to store class prediction into.  Will be set to length `data.n_cols`. |

***Note:*** different types can be used for `data` and `point` (e.g.
`arma::fmat`, `arma::sp_mat`, `arma::sp_vec`, etc.).  However, the element type
that is used should be the same type that was used for training.

### Other Functionality

 * `WeakLearner(size_t i)` returns a `WeakLearnerType` indicating which weak learner 
    is being used. 

 * `gb.NumWeakLearners()` returns a `size_t` indicating how many weak learners are being
    used by the model.

 * `gb.NumClasses()` returns a `size_t` indicating the number of classes the
   model was trained on.

 * `gb.SetNumWeakLearners(size_t x)` sets the number of weak learners to `x` explicitly.

For complete functionality, the [source
code](/src/mlpack/methods/grad_boosting/grad_boosting.hpp) can be consulted.
Each method is fully documented.

### Simple Examples

See also the [simple usage example](#simple-usage-example) for a trivial use of
`GradBoosting`.

---

> Train a gradient boosting model on mixed categorical data and save it:

Reviewer: I missed this on the first review: `GradBoosting` does not support mixed categorical data?  This example is actually insidiously wrong: it loads mixed categorical data, but then the `info` object is *ignored* when we call `Train()`.  So what actually happens in this example is that we load mixed categorical data, but then train a model, treating every dimension as a numeric type.  This is really misleading, and is a great example of why you cannot simply copy examples from other methods and adapt without thinking.

Either we should add support for categorical data in this PR, or, we should adapt this example to use a numerical-only dataset.

Author: Sorry about that, it was running on my system, so I didn't realize this. I'll add categorical support ASAP

Reviewer: Yeah, it will run just fine (and probably produce a good model too given the nature of the covertype dataset), but it subtly doesn't do what it says the example does.  Sounds good about the categorical support, thanks! :+1:

## `GradBoosting`

The `GradBoosting` class implements a Gradient Boosting model that supports
numerical and categorical features. The class offers several template 
parameters and several runtime options that can be used to control the 
model.

Gradient Boosting is a very powerful ensemble algorithm used for both 
classification and regression tasks. It utilizes a series of weak learners 

> (specifically [decision tree regressors](decision_tree_regressor.md)) to arrive closer and closer to a targeted label. Each 
> weak learner is trained on the error of the previous learner, thereby 
> reducing the error of the overall model.

Reviewer: ```suggestion
(specifically [decision tree regressors](decision_tree_regressor.md)), each
trained on the error of the previous learner.  This allows the ensemble to
produce accurate predictions.
```
I think this flows a little better, let me know what you think.

#### Simple usage example:

```c++
// Train a gradient boosting model on random numeric data and predict labels 
// on test data: All data and labels are uniform random; 10 dimensional data, 
// 5 classes.
// Replace with a data::Load() call or similar for a real application.
arma::mat dataset(10, 1000, arma::fill::randu); // 1000 points.
arma::Row<size_t> labels =
    arma::randi<arma::Row<size_t>>(1000, arma::distr_param(0, 4));
arma::mat testDataset(10, 500, arma::fill::randu); // 500 test points.
size_t numWeakLearners = 10;
size_t numClasses = 5;

> mlpack::GradBoosting gb;               // Step 1: create model.
> gb.Train(dataset, labels, numClasses, numWeakLearners);          // Step 2: train model.
> arma::Row<size_t> predictions;
> gb.Classify(testDataset, predictions); // Step 3: classify points.

Reviewer: ```suggestion
mlpack::GradBoosting gb;                                // Step 1: create model.
gb.Train(dataset, labels, numClasses, numWeakLearners); // Step 2: train model.
arma::Row<size_t> predictions;
gb.Classify(testDataset, predictions);                  // Step 3: classify points.
```

Having the comments aligned here can be helpful for readability.

    << "2." << std::endl;
```
<p style="text-align: center; font-size: 85%"><a href="#simple-examples">More examples...</a></p>

#### Quick links:

 * [Constructors](#constructors): create `GradientBoosting` objects.
 * [`Train()`](#training): train model.
 * [`Classify()`](#classification): classify with a trained model.
 * [Other functionality](#other-functionality) for loading, saving, and
   inspecting.
 * [Examples](#simple-examples) of simple usage and links to detailed example
   projects.

#### See also:

 * [`AdaBoost`](adaboost.md)
 * [`DecisionTree`](decision_tree.md)
 * [`DecisionTreeRegressor`](decision_tree_regressor.md)
 * [Random forests](random_forest.md)
 * [mlpack classifiers](../../index.md#classification-algorithms)
 * [Gradient Boosting on Wikipedia](https://en.wikipedia.org/wiki/Gradient_boosting)

### Constructors

 * `gb = GradBoosting()`
   - Initialize model without training.
   - You will need to call [`Train()`](#training) later to train the model before
     calling [`Classify()`](#classification).

---

 * `gb = GradBoosting(data, datasetInfo, labels, numClasses, numWeakLearners)`

>    - Train on numerical and/or categorical data using default arguments for the weak learners (Decision Trees).

Reviewer: ```suggestion
   - Train on numerical and/or categorical data using default arguments for the
     weak learners ([`DecisionTreeRegressor`s](decision_tree_regressor.md)).
```

---

 * `gb = GradBoosting(data, datasetInfo, labels, numClasses, numWeakLearners, minimumLeafSize, minimumGainSplit, maximumDepth)`
   - Train on numerical and/or categorical data, using the given hyperparameters.

> 

Reviewer: In these constructors, I would suggest laying them out like the `RandomForest` constructors, like this:

```
 * gb = GradBoosting()
   - default constructor section
   
---

 * gb = GradBoosting(numerical only data arguments)
 * gb = GradBoosting(numerical only data arguments plus hyperparameters)
   - numerical data constructor section
   
 ---
 
 * gb = GradBoosting(mixed categorical data arguments)
 * gb = GradBoosting(mixed categorical data arguments plus hyperparameters)
   - mixed categorical data constructor section 
```

This is primarily for consistency between different algorithms (since most of them have their constructors split up in this way).

---

#### Constructor Parameters:

| **name** | **type** | **description** | **default** |
|----------|----------|-----------------|-------------|
| `data` | [`arma::mat`](../matrices.md) | [Column-major](../matrices.md#representing-data-in-mlpack) training matrix. | _(N/A)_ |
| `datasetInfo` | [`data::DatasetInfo`](https://github.com/mlpack/mlpack/blob/master/doc/user/load_save.md#loading-categorical-data) | Dataset information, specifying type information for each dimension. | _(N/A)_ |
| `labels` | [`arma::Row<size_t>`](../matrices.md) | Training labels, [between `0` and `numClasses - 1`](../load_save.md#normalizing-labels) (inclusive).  Should have length `data.n_cols`.  | _(N/A)_ |
| `numClasses` | `size_t` | Number of classes in the dataset. | _(N/A)_ |

> | `numWeakLearners` | `size_t` | Number of weak learner models to be used. | _(N/A)_ |

Reviewer: I think it would be nice to specify a default argument here; after all, `RandomForest` and `AdaBoost` have defaults.

| `minimumLeafSize` | `size_t` | Minimum leaf size for weak learner decision tree. | `10` |

> | `minimumGainSplit` | `double` | Minimum gain split for weak learner decision tree | 1e-7 |
> | `maximumDepth` | `size_t` | Maximum tree depth for weak learner decision tree | 2 |

Reviewer: ```suggestion
| `minimumGainSplit` | `double` | Minimum gain split for weak learner decision tree | `1e-7` |
| `maximumDepth` | `size_t` | Maximum tree depth for weak learner decision tree | `2` |
```
A little formatting fix.

***Note:*** different types can be used for `data` (e.g.,

> `arma::fmat`, `arma::sp_mat`). 

Reviewer: Is the class tested on sparse matrices?  (I think it should work but it would be good to test it for sure.)

### Training

If training is not done as part of the constructor call, it can be done with one
of the following versions of the `Train()` member function:

 * `gb.Train(data, datasetInfo, labels, numClasses, numWeakLearners)`
   - Train on numerical data without using any decision tree arguments.

---

 * `gb.Train(data, datasetInfo, labels, numClasses, numWeakLearners, minimumLeafSize, minimumGainSplit, maximumDepth)`

>    - Train on numerical data using weak learner model's arguments `minimumLeafSize`, `minimumGainSplit` and `maximumDepth`.

Reviewer: ```suggestion
   - Train on mixed categorical data using the `minimumLeafSize`, `minimumGainSplit`, and `maximumDepth` options to control the behavior of the weak learners ([`DecisionTreeRegressor`s](decision_tree_regressor.md)).
```

Also: I would consider the names `minLeafSize`, `minGainSplit`, and `maxDepth` to save space.  When the arguments get too long, they render really badly.  (You can build the documentation locally with `./scripts/build-docs.sh` from the root of the repository, then point a browser at `doc/html/index.html` to see what I mean.)

One more comment about this section: I'd split it into two "sets" of `Train()` functions: one for numerical-only data (e.g. no `datasetInfo`) and one for mixed categorical data (e.g. with a `datasetInfo`).  I think in many of the other classes I called that `info` just to save space.

---

Types of each argument are the same as in the table for constructors
[above](#constructor-parameters).

***Note:***: Training is not incremental.  A second call to `Train()` will retrain the
model from scratch.

### Classification

Once a `GradBoosting` object is trained, the `Classify()` member function can be used
to make class predictions for new data.

 * `size_t predictedClass = gb.Classify(point)`
    - ***(Single-point)***
    - Classify a single point, returning the predicted class.

---

 * `gb.Classify(point, prediction)`
    - ***(Single-point)***
    - Classify a single point and compute class probabilities.
    - The predicted class is stored in `prediction`.

---

 * `gb.Classify(data, predictedLabels)`
    - ***(Multi-point)***
    - Classify a set of points.

>     - The prediction for data point `i` can be accessed with `predictedLabels[i]`.

Reviewer: Is there a multi-point version that will return class probabilities too?  If not we should add one for consistency.

---

#### Classification Parameters:

| **usage** | **name** | **type** | **description** |
|-----------|----------|----------|-----------------|
| _single-point_ | `point` | [`arma::vec`](../matrices.md) | Single point for classification. |
| _single-point_ | `prediction` | `size_t&` | `size_t` to store class prediction into. |
||||
| _multi-point_ | `data` | [`arma::mat`](../matrices.md) | Set of [column-major](../matrices.md#representing-data-in-mlpack) points for classification. |
| _multi-point_ | `predictedLabels` | [`arma::Row<size_t>&`](../matrices.md) | Vector of `size_t`s to store class prediction into.  Will be set to length `data.n_cols`. |

***Note:*** different types can be used for `data` and `point` (e.g.
`arma::fmat`, `arma::sp_mat`, `arma::sp_vec`, etc.).  However, the element type
that is used should be the same type that was used for training.

### Other Functionality

>  * `WeakLearner(size_t i)` returns a `DecisionTreeRegressor&` indicating which weak learner 

Reviewer: ```suggestion
 * `WeakLearner(size_t i)` returns a [`DecisionTreeRegressor&`](decision_tree_regressor.md) indicating which weak learner 
```
Could be nice to add a convenient link.

    is being used. 

 * `gb.NumWeakLearners()` returns a `size_t` indicating how many weak learners are being
    used by the model.

 * `gb.NumClasses()` returns a `size_t` indicating the number of classes the
   model was trained on.

>  * `gb.SetNumWeakLearners(size_t x)` sets the number of weak learners to `x` explicitly.

Reviewer: For consistency with other classes, I would rename this to `gb.NumWeakLearners(const size_t)` in the code, and then document it here with `NumWeakLearners()`:

```
 * `gb.NumWeakLearners()` returns a `size_t` indicating how many weak learners are being
    used by the model.
    - `gb.NumWeakLearners(x)` will set the number of weak learners that will be used the *next time* `Train()` is called to `x`.
```

### Simple Examples

See also the [simple usage example](#simple-usage-example) for a trivial use of
`GradBoosting`.

---

Train a gradient boosting model on mixed categorical data and save it:

```c++
// Load a categorical dataset.
arma::mat dataset;
mlpack::data::DatasetInfo info;
// See https://datasets.mlpack.org/covertype.train.arff.
mlpack::data::Load("covertype.train.arff", dataset, info, true);

arma::Row<size_t> labels;
// See https://datasets.mlpack.org/covertype.train.labels.csv.
mlpack::data::Load("covertype.train.labels.csv", labels, true);

// Create the model object.
mlpack::GradBoosting gb;
// Train on the given dataset, specifying the number of weak learners as 5.
gb.Train(dataset, info, labels, 7 /* classes */, 5 /* number of weak learners */);

// Load categorical test data.
arma::mat testDataset;
// See https://datasets.mlpack.org/covertype.test.arff.
mlpack::data::Load("covertype.test.arff", testDataset, true);

// Predict class of first test point.
const size_t firstPrediction = gb.Classify(testDataset.col(0));
std::cout << "Predicted class of first test point is " << firstPrediction << "."
    << std::endl;

// Save the model to `gb.bin`.

> mlpack::data::Save("gb.bin", "gb", gb);

Reviewer: I've tried in the examples I've written to keep each example concise and demonstrating only one thing (so much as is possible, at least).  So I would suggest splitting into a second example here, where you could:

 * train a `GradBoosting` on the covertype dataset, but passing arguments to the constructor
 * save to disk

```

---

Load a model and print some information about it.

```c++
mlpack::GradBoosting gb;
// This call assumes a gb called "gb" has already been saved to `gb.bin`
// with `data::Save()`.
mlpack::data::Load("gb.bin", "gb", gb, true);

> std::cout << "The number of weak learners being used by the model is "
> << gb.NumWeakLearners() << "." << std::endl;

Reviewer: ```suggestion
std::cout << "The number of weak learners being used by the model is "
    << gb.NumWeakLearners() << "." << std::endl;
```

Reviewer: If you can print more information through some of the other accessors (like maybe a couple things from the first weak learner) that would be nice too.

> ## `GradBoosting`

Author: Implemented this :+1:

The `GradBoosting` class implements a Gradient Boosting model that supports

> numerical and categorical features. The class offers several template 

Reviewer: Does this actually support categorical features?  I don't think it does currently.

arma::mat dataset(10, 1000, arma::fill::randu); // 1000 points.
arma::Row<size_t> labels =
    arma::randi<arma::Row<size_t>>(1000, arma::distr_param(0, 4));
arma::mat testDataset(10, 500, arma::fill::randu); // 500 test points.
size_t numModels = 10;
size_t numClasses = 5;

mlpack::GradBoosting gb;               // Step 1: create model.
gb.Train(dataset, labels, numClasses, numModels);          // Step 2: train model.
arma::Row<size_t> predictions;
gb.Classify(testDataset, predictions); // Step 3: classify points.

// Print some information about the test predictions.
std::cout << arma::accu(predictions == 2) << " test points classified as class "
    << "2." << std::endl;
```
<p style="text-align: center; font-size: 85%"><a href="#simple-examples">More examples...</a></p>

#### Quick links:

 * [Constructors](#constructors): create `GradientBoosting` objects.
 * [`Train()`](#training): train model.
 * [`Classify()`](#classification): classify with a trained model.
 * [Other functionality](#other-functionality) for loading, saving, and
   inspecting.
 * [Examples](#simple-examples) of simple usage and links to detailed example
   projects.

#### See also:

>  * [`DecisionTree`](decision_tree.md)

Author: Implemented this :+1:

---

 * `gb = GradBoosting(data, labels, numClasses, numModels, minimumLeafSize, minimumGainSplit, maximumDepth)`
   - Train on numerical-only data, entering the weak learner arguments `minimumLeafSize`, `minimumGainSplit` and `maximumDepth`.

---

#### Constructor Parameters:

| **name** | **type** | **description** | **default** |
|----------|----------|-----------------|-------------|
| `data` | [`arma::mat`](../matrices.md) | [Column-major](../matrices.md#representing-data-in-mlpack) training matrix. | _(N/A)_ |
| `labels` | [`arma::Row<size_t>`](../matrices.md) | Training labels. [between `0` and `numClasses - 1`](../load_save.md#normalizing-labels) (inclusive).  Should have length `data.n_cols`.  | _(N/A)_ |
| `numClasses` | `size_t` | Number of classes in the dataset. | _(N/A)_ |
| `numModels` | `size_t` | Number of weak learner models to be used. | _(N/A)_ |
| `minimumLeafSize` | `size_t` | Minimum leaf size for weak learner decision tree | 10 |
| `minimumGainSplit` | `double` | Minimum gain split for weak learner decision tree | 1e-7 |
| `maximumDepth` | `size_t` | Maximum tree depth for weak learner decision tree | 2 |

>  * Number of labels in the data must be less than or equal to numClasses.

Author: Implemented this :+1:

    - The predicted class is stored in `prediction`.

---

 * `gb.Classify(data, predictedLabels)`
    - ***(Multi-point)***
    - Classify a set of points.
    - The prediction for data point `i` can be accessed with `predictedLabels[i]`.

---

#### Classification Parameters:

| **usage** | **name** | **type** | **description** |
|-----------|----------|----------|-----------------|
| _single-point_ | `point` | [`arma::vec`](../matrices.md) | Single point for classification. |
| _single-point_ | `prediction` | `size_t&` | `size_t` to store class prediction into. |
||||
| _multi-point_ | `data` | [`arma::mat`](../matrices.md) | Set of [column-major](../matrices.md#representing-data-in-mlpack) points for classification. |
| _multi-point_ | `predictedLabels` | [`arma::Row<size_t>&`](../matrices.md) | Vector of `size_t`s to store class prediction into.  Will be set to length `data.n_cols`. |

***Note:*** different types can be used for `data` and `point` (e.g.
`arma::fmat`, `arma::sp_mat`, `arma::sp_vec`, etc.).  However, the element type
that is used should be the same type that was used for training.

### Other Functionality

>  * `WeakLearner(size_t i)` returns a `WeakLearnerType` indicating which weak learner 

Author: For now it's only `DecisionTreeRegressor`, I'll make it a template function in a later PR, so I've left it as `WeakLearnerType` for now.

>  * `WeakLearner(size_t i)` returns a `WeakLearnerType` indicating which weak learner 

Reviewer: Adding templated support would be awesome, but at least in the documentation for now we haven't defined anywhere what `WeakLearnerType` is.  So I would suggest changing to:

```suggestion
 * `WeakLearner(size_t i)` returns a `DecisionTreeRegressor&` indicating which weak learner 
```

Later when templated support for other weak learners is added, we can adapt this accordingly.

    is being used. 

 * `gb.NumModels()` returns a `size_t` indicating how many weak learners are being
    used by the model.

 * `gb.NumClasses()` returns a `size_t` indicating the number of classes the
   model was trained on.

>  * `gb.SetNumModels(size_t x)` set the number of weak learners to `x` explicitly.

Reviewer: I think that this comment still needs to be handled. :+1: (I saw you resolved it, but it seems like the bullet points aren't addressed?  Maybe I overlooked something, let me know if so.)

arma::Row<size_t> labels;
// See https://datasets.mlpack.org/covertype.train.labels.csv.
mlpack::data::Load("covertype.train.labels.csv", labels, true);

// Create the model object.
mlpack::GradBoosting gb;
// Train on the given dataset, specifying number of weak learners at 5.
gb.Train(dataset, labels, 7 /* classes */, 5 /* number of weak learners */);

// Load categorical test data.
arma::mat testDataset;
// See https://datasets.mlpack.org/covertype.test.arff.
mlpack::data::Load("covertype.test.arff", testDataset, info, true);

// Predict class of first test point.
const size_t firstPrediction = gb.Classify(testDataset.col(0));
std::cout << "Predicted class of first test point is " << firstPrediction << "."
    << std::endl;

// Save the model to `gb.bin`.
mlpack::data::Save("gb.bin", "gb", gb);
```

---

Load a model and print some information about it.

```c++
mlpack::GradBoosting gb;
// This call assumes a gb called "gb" has already been saved to `gb.bin`
// with `data::Save()`.
mlpack::data::Load("gb.bin", "gb", gb, true);

std::cout << "The number of weak learners being used by the model is "
<< gb.NumModels() << "." << std::endl;

```

> ---

Reviewer: Thanks for working on the examples.  Still the ideas above would be worth adding when you have time.  One method I don't see used (and I know it is easy, but still, having an actual demonstration is always helpful) is the batch prediction function.  One of the things I've done in other documentation is to use the batch prediction function plus a quick calculation to compute accuracy on the entire test set, or something like this.
